utils.ts

```diff
@@ -18,54 +22,82 @@ export function u32_merge_bits(high: u16, low: u16): u32 {
   return high as u32 << 16 | (low as u32);
 }
 
-export function check_encoded_allocation(encoded_allocation: u32): u32 {
-  
-  let offset = u32_high_bits(encoded_allocation);
-  let length = u32_low_bits(encoded_allocation);
 
+export function check_encoded_allocation(encoded_allocation: u32): ErrorCode {
+  let offset: u16 = u32_high_bits(encoded_allocation);
+  let length: u16 = u32_low_bits(encoded_allocation);
   if (length == 0) {
-    // error
-    return 1 << 16;
+    return offset as ErrorCode;
   }
-
   // switch to u32 from u16
   let u32offset: u32 = offset;
   let u32length: u32 = length;
   let max: u32 = u16.MAX_VALUE;
   if ((u32offset + u32length) > max) {
-    // ErrorPageOverflow
-    return 3 << 16;
+    return ErrorCode.PageOverflowError;
   }
-
-  // 0 is success
-  return 0;
+  return ErrorCode.Success;
 }
 
+
 // writes string to memory, then returns encoded allocation ref
 export function serialize(val: string): u32 {
   let dataLength = val.length;
   // each char takes two bytes, encoded
   let ptr = memory.allocate(dataLength << 1);
   //checkMem();
   for (let i = 0; i < dataLength; ++i) {
-    store<u8>(ptr + i, val.charCodeAt(i));
+    store<u16>(ptr + i, val.charCodeAt(i));
   }
   let encoded_allocation = u32_merge_bits(ptr as u16, dataLength as u16);
   return encoded_allocation;
 }
 
+
+// reads a string into a new memory allocation that uses the format for asm
 export function deserialize(encoded_allocation: u32): string {
   let offset = u32_high_bits(encoded_allocation);
   let length = u32_low_bits(encoded_allocation);
-  let res: string = "";
+  let res: string = allocateUnsafe(length);
+
+  // TODO: figure out how to do this in a single copy. Need to change boundaries on characters
   for (let i: u16 = 0; i < length; i++) {
-    let charCode = load<u8>(offset + i);
-    res += String.fromCharCode(charCode);
+    memory.copy(
+      changetype<usize>(res) + HEADER_SIZE + (i<<1),
+      changetype<usize>(offset) + i,
+      1
+    );
   }
-  memory.free(offset);
   return res;
 }
 
+
 export function free(ptr: u32): void {
   memory.free(ptr);
+}
+
+
+export function errorCodeToString(code: ErrorCode): string {
```

Comment:
We'll prbly want to create some kind of util that let's folks do error checking of the strings too, to see if what they got back was an error, or a good result